### ISSUE TITLE
Are patches and extensions now welcome?

### ISSUE THREAD

**Issue by saaadel** [dayllenger/beamui#10]
Because at the moment it is WIP and there is no fixed structure of the library, are you currently considering proposals for expanding the library with ready-made code (PR), or only through preliminary approval of ideas before starting to somehow expand the library?

**Comment by dayllenger** [comment]
It's better to discuss before doing PR, if it's not a bugfix.

**Comment by saaadel** [comment]
At first I want following:
1. More layouts: for example, 
- FlowLayout (nodes in the horizontal rows according to the available horizontal spaces, and wraps the nodes to the next line if the horizontal space is less than the total width of the nodes - works like inline-blocks in the HTML rows/lines); 
- BorderLayout (common top, bottom, left, right, and center areas for widgets - single widget per an area); 

2. To add "skin" CSS property for every widget class - separate D class name, that have an extracted logic - how to draw the widget, and how to calculate measured sizes. It should be an optional property and have a predefined value for every widget. Rationale: Special implementations from users, but keeping all other logic same separate in widget class; Possible a special default implementations for some platforms. (to support the latter, I would prefer that we have common version tokens support in media queries)

3. Add some more widgets

**Comment by saaadel** [comment]
4. RTL support - NodeOrientation: LEFT_TO_RIGHT/ RIGHT_TO_LEFT

**Comment by dayllenger** [comment]
> FlowLayout

CSS flexbox is better (so FlexLayout).
> BorderLayout

Yes, this is pretty easy to implement. `alignment` for alignment, `rowGap` and `columnGap` for space between items, `display` set to "border" value. Partly it can be emulated now with "free" layout and `top`, `right`, `bottom`, `left` properties.

> "skin"

I need examples. Assume I have a simple ImageWidget and a somewhat complex EditLine. What can I do with them using `skin`?

> Add some more widgets

I wanted to add color picker, font picker, calendar, date-time editors, spin boxes (the last two are kinda frozen because of lacking support of editor validators and stuff).

> RTL

I know little about RTL. Implementing various writing modes and right-to-left scripts is a massive task. If we don't support RTL for text, do we need it for layouts? Also, it's better to use `direction` CSS property, as they can inherit.

**Comment by saaadel** [comment]
> What can I do with them using skin?

Basically the idea is that, now we can only change the style of the widget, but it will consist of all the same elements as before. Skin can replace the render logic of the widget, but all other components will see it as a standard widget with all data on its properties.
I can't immediately come up with a solution for the components you specified, but, for example, for tabs using custom skins User can place some tabs on left side, some on right side, and some in the middle - all of them on top for example:
```
A/B/C              D/E/F                X/Y/Z
(               TABS CONTENT HERE        )
```
or better example - there can be too many tabs, and different skins can render them in different ways:

1)

```
< A/B/C.................................................. >
(               TABS CONTENT HERE        )
```
left/right buttons to scroll tabs

2)

```
A/B/C.................................................. [v]
(               TABS CONTENT HERE        )
```
[v] dropdown arrow with full list of the tabs - just click to select needed one

3)

```
A/B/C.........................................J/K/L
M/N/O.........................................X/Y/Z
(               TABS CONTENT HERE        )
```
show tabs on multiple rows

>  editor validators

I have some ideas here from different UI libraries. Let me take some time to review it all.

> RTL

hm, without text it does not have sense. I need review it for text before...

**Comment by saaadel** [comment]
5. combining diacritics characters as single glyphs support - required! now it's broken

**Comment by dayllenger** [comment]
> for tabs

So currently tab placement logic [is concentrated here](https://github.com/dayllenger/beamui/blob/84ed04d75ee618d178f721f5d7243135d49639de/src/beamui/widgets/tabs.d#L478) in `measure`, `layout`, and `drawContent`, plus `sortedItems` and `tabAlignment` above. So I guess it's obvious that those skin objects 1) apply to a specific type of widget, 2) can look at its properties, 3) should be accessible somehow by the user to adjust their specific properties.

It doesn't look very complex to implement, but it needs more examples of how it can be useful in other places. In the example above, I can basically do it with inheritance.

**Comment by saaadel** [comment]
Yes, it just adds more flexible rendering control. In any case, I want come back to this later, and with some implementation example, because this can be replaced by a bunch of inheritance from the basic components (in most cases) - not so flexible, but already a working solution.

In the meantime, for now, I'll start implementing RTL for text, and to fix diacritics characters

**Comment by dayllenger** [comment]
Main things which render text are [SimpleText](https://dayllenger.github.io/beamui/beamui/text/simple.html) for text without inner formatting and [TextLine](https://dayllenger.github.io/beamui/beamui/text/line.html) for markup.

Formatted text is not documented yet, I can give some examples and explanations of how it works if you need.

**Comment by aberba** [comment]
I like how things are going especially the flexbox and grid design
